File: bba-server-rs/src/services/convention_service.rs
use regex::Regex;
use tracing::{info, warn};

/// Fetches convention cards (.bbsa) and scenario files (.pbs) from GitHub.
pub struct ConventionService {
    client: reqwest::Client,
    github_raw_base_url: String,
    default_ns_card: String,
    default_ew_card: String,
}

impl ConventionService {
    pub fn new(github_raw_base_url: &str, default_ns_card: &str, default_ew_card: &str) -> Self {
        Self {
            client: reqwest::Client::new(),
            github_raw_base_url: github_raw_base_url.to_string(),
            default_ns_card: default_ns_card.to_string(),
            default_ew_card: default_ew_card.to_string(),
        }
    }

    pub fn default_ns(&self) -> &str {
        &self.default_ns_card
    }

    pub fn default_ew(&self) -> &str {
        &self.default_ew_card
    }

    /// Fetch a .bbsa convention card from GitHub and return its content.
    pub async fn get_bbsa_content(&self, card_name: &str) -> Result<String, String> {
        let url = format!("{}/bbsa/{}.bbsa", self.github_raw_base_url, card_name);
        info!("Fetching convention card from GitHub: {}", url);

        let resp = self
            .client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("Failed to fetch convention card: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "Convention card not found on GitHub: {} (HTTP {})",
                card_name,
                resp.status()
            ));
        }

        resp.text()
            .await
            .map_err(|e| format!("Failed to read convention card: {}", e))
    }

    /// Get conventions for a scenario by fetching its .pbs file from GitHub.
    pub async fn get_conventions_for_scenario(
        &self,
        scenario: &str,
    ) -> (String, String) {
        match self.get_convention_cards_from_pbs(scenario).await {
            (Some(ns), Some(ew)) => (ns, ew),
            (Some(ns), None) => (ns, self.default_ew_card.clone()),
            (None, Some(ew)) => (self.default_ns_card.clone(), ew),
            (None, None) => (self.default_ns_card.clone(), self.default_ew_card.clone()),
        }
    }

    async fn get_convention_cards_from_pbs(&self, scenario: &str) -> (Option<String>, Option<String>) {
        let url = format!("{}/pbs-release/{}.pbs", self.github_raw_base_url, scenario);

        let resp = match self.client.get(&url).send().await {
            Ok(r) => r,
            Err(e) => {
                warn!("Error fetching PBS file for {}: {}", scenario, e);
                return (None, None);
            }
        };

        if !resp.status().is_success() {
            warn!("PBS file not found for {} (HTTP {})", scenario, resp.status());
            return (None, None);
        }

        let content = match resp.text().await {
            Ok(c) => c,
            Err(_) => return (None, None),
        };

        // Try new format: convention-card-ns and convention-card-ew
        let ns_re = Regex::new(r"(?mi)^convention-card-ns:\s*(.+)$").unwrap();
        let ew_re = Regex::new(r"(?mi)^convention-card-ew:\s*(.+)$").unwrap();

        let ns = ns_re.captures(&content).map(|c| c[1].trim().to_string());
        let ew = ew_re.captures(&content).map(|c| c[1].trim().to_string());

        if ns.is_some() || ew.is_some() {
            info!(
                "Found convention cards NS='{}', EW='{}' for scenario '{}'",
                ns.as_deref().unwrap_or("(default)"),
                ew.as_deref().unwrap_or("(default)"),
                scenario
            );
            return (ns, ew);
        }

        // Fall back to old format: convention-card (NS only)
        let old_re = Regex::new(r"(?mi)^#?\s*convention-card:\s*(.+)$").unwrap();
        if let Some(caps) = old_re.captures(&content) {
            let value = caps[1].trim().to_string();
            if !value.is_empty() {
                info!(
                    "Found convention card '{}' (old format) for scenario '{}'",
                    value, scenario
                );
                return (Some(value), None);
            }
        }

        info!("No convention card specified for scenario '{}', using defaults", scenario);
        (None, None)
    }

    /// Get the list of available scenarios from GitHub.
    pub async fn get_scenario_list(&self) -> Result<Vec<String>, String> {
        let url =
            "https://api.github.com/repos/ADavidBailey/Practice-Bidding-Scenarios/contents/pbs-release";

        let resp = self
            .client
            .get(url)
            .header("User-Agent", "BBA-Server")
            .send()
            .await
            .map_err(|e| format!("Failed to fetch scenario list: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("GitHub API returned {}", resp.status()));
        }

        let items: Vec<serde_json::Value> = resp
            .json()
            .await
            .map_err(|e| format!("Failed to parse GitHub response: {}", e))?;

        let mut scenarios: Vec<String> = items
            .iter()
            .filter_map(|item| {
                let name = item["name"].as_str()?;
                if name.ends_with(".pbs") {
                    Some(name.trim_end_matches(".pbs").to_string())
                } else {
                    None
                }
            })
            .collect();

        scenarios.sort();
        Ok(scenarios)
    }
}

File: legacy/bba-server-cs/Services/ConventionService.cs
using System.Text.RegularExpressions;
using BbaServer.Models;

namespace BbaServer.Services;

/// <summary>
/// Service for looking up convention cards from GitHub.
/// Fetches .bbsa files from raw.githubusercontent.com and .pbs files for scenario convention lookups.
/// </summary>
public class ConventionService
{
    private readonly ILogger<ConventionService> _logger;
    private readonly HttpClient _httpClient = new();
    private readonly string _defaultNsCard;
    private readonly string _defaultEwCard;
    private readonly string _githubRawBaseUrl;

    public ConventionService(ILogger<ConventionService> logger, IConfiguration configuration)
    {
        _logger = logger;

        _defaultNsCard = configuration["Pbs:DefaultNsCard"] ?? "21GF-DEFAULT";
        _defaultEwCard = configuration["Pbs:DefaultEwCard"] ?? "21GF-GIB";
        _githubRawBaseUrl = configuration["GitHub:RawBaseUrl"]
            ?? "https://raw.githubusercontent.com/ADavidBailey/Practice-Bidding-Scenarios/main";
    }

    /// <summary>
    /// Get convention cards for a scenario by fetching its .pbs file from GitHub.
    /// </summary>
    public async Task<ConventionCards> GetConventionsForScenarioAsync(string scenario)
    {
        var (nsCard, ewCard) = await GetConventionCardsFromPbs(scenario);
        return new ConventionCards
        {
            Ns = nsCard ?? _defaultNsCard,
            Ew = ewCard ?? _defaultEwCard
        };
    }

    /// <summary>
    /// Fetch a .bbsa convention card file from GitHub and return its lines.
    /// </summary>
    public async Task<string[]> GetBbsaContentAsync(string cardName)
    {
        var url = $"{_githubRawBaseUrl}/bbsa/{cardName}.bbsa";
        _logger.LogInformation("Fetching convention card from GitHub: {Url}", url);

        var response = await _httpClient.GetAsync(url);
        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"Convention card not found on GitHub: {cardName} (HTTP {(int)response.StatusCode})");
        }

        var content = await response.Content.ReadAsStringAsync();
        return content.Split('\n');
    }

    /// <summary>
    /// Fetch a scenario's .pbs file from GitHub and extract convention card properties.
    /// Supports both new format (convention-card-ns/ew) and old format (convention-card).
    /// </summary>
    private async Task<(string? ns, string? ew)> GetConventionCardsFromPbs(string scenario)
    {
        var url = $"{_githubRawBaseUrl}/pbs-release/{scenario}.pbs";

        try
        {
            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                _logger.LogWarning("PBS file not found on GitHub: {Scenario} (HTTP {StatusCode})", scenario, (int)response.StatusCode);
                return (null, null);
            }

            var content = await response.Content.ReadAsStringAsync();

            // Try new format first: convention-card-ns and convention-card-ew
            var nsPattern = new Regex(@"^convention-card-ns:\s*(.+)$", RegexOptions.IgnoreCase | RegexOptions.Multiline);
            var ewPattern = new Regex(@"^convention-card-ew:\s*(.+)$", RegexOptions.IgnoreCase | RegexOptions.Multiline);
            var nsMatch = nsPattern.Match(content);
            var ewMatch = ewPattern.Match(content);

            if (nsMatch.Success || ewMatch.Success)
            {
                var ns = nsMatch.Success ? nsMatch.Groups[1].Value.Trim() : null;
                var ew = ewMatch.Success ? ewMatch.Groups[1].Value.Trim() : null;
                _logger.LogInformation("Found convention cards NS='{Ns}', EW='{Ew}' for scenario '{Scenario}'",
                    ns ?? "(default)", ew ?? "(default)", scenario);
                return (ns, ew);
            }

            // Fall back to old format: convention-card (NS only)
            var oldPattern = new Regex(@"^#?\s*convention-card:\s*(.+)$", RegexOptions.IgnoreCase | RegexOptions.Multiline);
            var oldMatch = oldPattern.Match(content);
            if (oldMatch.Success)
            {
                var value = oldMatch.Groups[1].Value.Trim();
                if (!string.IsNullOrEmpty(value))
                {
                    _logger.LogInformation("Found convention card '{Card}' (old format) for scenario '{Scenario}'", value, scenario);
                    return (value, null);
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching PBS file from GitHub for scenario: {Scenario}", scenario);
        }

        _logger.LogInformation("No convention card specified for scenario '{Scenario}', using defaults", scenario);
        return (null, null);
    }

    /// <summary>
    /// Get the list of available scenarios from GitHub's pbs-release folder.
    /// </summary>
    public async Task<List<string>> GetScenarioListAsync()
    {
        var url = "https://api.github.com/repos/ADavidBailey/Practice-Bidding-Scenarios/contents/pbs-release";
        _logger.LogInformation("Fetching scenario list from GitHub API");

        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Add("User-Agent", "BBA-Server");

        var response = await _httpClient.SendAsync(request);
        if (!response.IsSuccessStatusCode)
        {
            _logger.LogWarning("Failed to fetch scenario list from GitHub (HTTP {StatusCode})", (int)response.StatusCode);
            return new List<string>();
        }

        var json = await response.Content.ReadAsStringAsync();
        var items = System.Text.Json.JsonSerializer.Deserialize<List<GitHubContentItem>>(json);
        if (items == null) return new List<string>();

        return items
            .Where(i => i.name?.EndsWith(".pbs") == true)
            .Select(i => Path.GetFileNameWithoutExtension(i.name!))
            .OrderBy(s => s)
            .ToList();
    }

    private class GitHubContentItem
    {
        public string? name { get; set; }
        public string? type { get; set; }
    }
}
